### PR TITLE
BeegFS Fix

### PR DESCRIPTION
**Pull request by brtknr**
Minor fixes to get beegfs working

### CODE REVIEW

File: tasks/main.yml [hunk 1]
```diff
@@ -20,6 +20,12 @@
   become: True
   with_items: "{{ beegfs_kmod_preload }}"
 
+- name: Generate /etc/hosts file
```
Comment (oneswig) [inline tasks/main.yml]: I think this is out of scope for BeeGFS and ought to be done elsewhere, for example in p3-appliances cluster_setup.

File: tasks/main.yml [hunk 2]
```diff
@@ -20,16 +20,32 @@
   become: True
   with_items: "{{ beegfs_kmod_preload }}"
 
-- name: Stop all beegfs services
+- name: Generate /etc/hosts file
+  template:
+    src: hosts.j2
+    dest: /etc/hosts
+  become: yes
```
Comment (markgoddard) [inline tasks/main.yml]: This is something that should be listed as a prerequisite and handled outside of the role.

File: tasks/main.yml [hunk 3]
```diff
@@ -28,8 +34,10 @@
 
 - name: Unmount beegfs data
   mount:
-    path: /data/beegfs/beegfs_data_0
-    state: unmounted
+    path: "{{ item }}"
+    state: absent
+  with_items:
+  - "{{ beegfs_path_oss }}"
```
Comment (oneswig) [inline tasks/main.yml]: I think it would be better to separate the "uncreate" of a BeeGFS deployment into a separate set of tasks.  If you conditionally include either "create.yml" or "destroy.yml" for example.

File: tasks/main.yml (hunk 2 above)
Comment (oneswig) [reply tasks/main.yml]: I added a comment in my review but it's not here - I think this is out of scope for BeeGFS and should be handled eg by cluster_setup in p3-appliances.

File: tasks/main.yml (hunk 3 above)
Comment (brtknr) [reply tasks/main.yml]: thats what I was thinking... using e.g. beegfs_state: "present" / "absent" to conditionally invoke "create.yml" and "destroy.yml" playbooks

**Comment by brtknr** [comment]
More idempotent than its ever been but still room for more idempotency! Ready for review @oneswig  @markgoddard :)

File: README.md [hunk 4]
```diff
@@ -1,32 +1,40 @@
...
+named in the inventory.
+
+Additionally, the following groups must be defined in the appliance
+inventory:
+
+- `cluster_beegfs_mgmt` (Management server - minimum one host)
+- `cluster_beegfs_mds` (Metadata storage server nodes)
+- `cluster_beegfs_oss` (Object storage server nodes)
+- `cluster_beegfs_client` (Clients of the BeeGFS storage cluster)
+- `cluster_beegfs_admon` (NOT IMPLEMENTED)
```
[the remark this answers is not in the document]
Comment (brtknr) [reply README.md]: Elaborate please...

File: README.md [hunk 5]
```diff
@@ -1,32 +1,40 @@
...
-    # - beegfs_mds (Metadata storage server nodes)
-    # - beegfs_oss (Object storage server nodes)
-    # - beegfs_client (Clients of the BeeGFS storage cluster)
     - hosts: cluster
       roles:
         - role: stackhpc.beegfs
           beegfs_enable:
             mgmt: "{{ inventory_hostname in groups['cluster_beegfs_mgmt'] }}"
-            admon: False
+            admon: no
```
[the remark this answers is not in the document]
Comment (brtknr) [reply README.md]: What is the best practice here?

File: defaults/main.yml [hunk 6]
```diff
@@ -2,16 +2,13 @@
...
 
 beegfs_path_client: "/mnt/beegfs"
 beegfs_path_mgmt: "/data/beegfs/beegfs_mgmtd"
 beegfs_path_meta: "/data/beegfs/beegfs_meta"
-beegfs_path_oss:
-  - "/data/beegfs/beegfs_data_0"
-  - "/data/beegfs/beegfs_data_1"
-  - "/data/beegfs/beegfs_data_2"
-  - "/data/beegfs/beegfs_data_3"
+beegfs_path_oss: "/data/beegfs/beegfs_oss"
```
[the remark this answers is not in the document]
Comment (brtknr) [reply defaults/main.yml]: Sure, this seemed like unnecessary repetition for the user... making it more similar to glusferfs role in implementation detail.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (markgoddard) [reply README.md]: The role does not reference those groups - only the example playbook in the README does. Roles in general should be reusable, and should not make any assumptions about the structure or naming of the inventory. This is what the `beegfs_enable` flags are for - they allow a playbook to specify which services a host should run.

File: README.md [hunk 7]
```diff
@@ -1,32 +1,100 @@
...
+
+    # {{ ansible_managed }}
+    127.0.0.1   localhost localhost.localdomain localhost4
+    localhost4.localdomain4
+    ::1         localhost localhost.localdomain localhost6
+    localhost6.localdomain6
+
+    {% for item in ansible_play_hosts %}
+    {% set short_name = item.split('.') %}
+    {{ hostvars[item]['ansible_host'] }}  {{ item }} {{ short_name[0] }}
```
[the remark this answers is not in the document]
Comment (brtknr) [reply README.md]: whatever the ilab equivalent is?

[the remark this answers is not in the document]
Comment (markgoddard) [reply README.md]: But more generally, in BeeGFS terms? Are the relevant IPs referenced anywhere in the BeeGFS config?

[the remark this answers is not in the document]
Comment (markgoddard) [reply README.md]: Ok, so do the hostnames in /etc/hosts need to map to the IP address used by the nodes when communicating with the configured management IP?

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (brtknr) [reply README.md]: I have addressed this, all toggles have the `bool` filter

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (brtknr) [reply README.md]: yup

[the remark this answers is not in the document]
Comment (markgoddard) [reply README.md]: This one still needs fixing - ansible_host might not be an IP. I think something like this would be sufficient here:

> Each node's hostname must resolve to the IP address used to reach the management host, as configured via `beegfs_host_mgmt`. This may be done via DNS or `/etc/hosts`.

File: README.md [hunk 8]
```diff
@@ -1,32 +1,41 @@
 # stackhpc.beegfs
```
Comment (markgoddard) [inline README.md]: Like the new README.

File: tasks/destroy.yml [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  become: yes
+  with_items:
+  - "{{ beegfs_dev_oss }}"
+  when:
+  - beegfs_enable.oss | bool
+  - beegfs_dev_oss is defined
+
+- name: Remove all beegfs folders
+  file:
+    path: /data/beegfs/
```
Comment (markgoddard) [inline tasks/destroy.yml]: Is this path always correct?

File: tasks/oss.yml [hunk 10]
```diff
@@ -3,58 +3,51 @@
...
-    - "{{ beegfs_path_oss }}"
-  - name: Format filesystem
+  - name: Attempt to format if the device is not mounted
+    vars:
+      path_dev_item: "{{ beegfs_path_dev }}/{{ item }}"
     filesystem:
-      dev: "{{ item }}"
+      dev: "{{ path_dev_item }}"
       fstype: "{{ beegfs_fstype }}"
-      force: "{{ beegfs_force_format }}"
```
Comment (markgoddard) [inline tasks/oss.yml]: The force has been removed here, what if the disk already has a file system on it?

File: tasks/rdma.yml [hunk 11]
```diff
@@ -9,17 +9,15 @@
...
     state: present
-  become: True
+  become: yes
   with_items:
   - "ib_umad"
   - "ib_uverbs"
   - "rdma_ucm"
-  when: False
-
+...
```
Comment (markgoddard) [inline tasks/rdma.yml]: I don't see this file being included anywhere.

File: tasks/oss.yml (hunk 10 above)
Comment (brtknr) [reply tasks/oss.yml]: Then the user ought to intervene no?

File: tasks/destroy.yml (hunk 9 above)
Comment (brtknr) [reply tasks/destroy.yml]: Perhaps better to iterate over the paths that we definitely know exist! This is only correct for the current configuration.

File: README.md (hunk 8 above)
Comment (brtknr) [reply README.md]: Thanks!

File: tasks/oss.yml (hunk 10 above)
Comment (markgoddard) [reply tasks/oss.yml]: Well yes, but removing the force option has removed a feature of the role, where if you set `beegfs_force_format` to true then it will deploy a file system on disks even if they already contain data. Another option would be to provide a mechanism for wiping devices prior to creating the cluster.

File: tasks/destroy.yml [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    name: "{{ item.name }}"
+    enabled: no
+    state: stopped
+  become: yes
+  when: item.enable | bool
+  with_items:
+  - { name: beegfs-mgmtd, enable: "{{ beegfs_enable.mgmt }}" }
+  - { name: beegfs-meta, enable: "{{ beegfs_enable.meta }}" }
+  - { name: beegfs-storage, enable: "{{ beegfs_enable.oss }}" }
+  - { name: beegfs-admon, enable: "{{ beegfs_enable.admon }}" }
```
Comment (markgoddard) [inline tasks/destroy.yml]: What about helperd and client?

File: tasks/oss.yml [hunk 13]
```diff
@@ -3,30 +3,48 @@
...
 
 - name: Ensures BeeGFS storage service is not running
   service:
-    name: "beegfs-storage"
-    state: "stopped"
-  become: True
+    name: "{{ item }}"
+    state: stopped
+  with_items:
+  - beegfs-client
```
Comment (markgoddard) [inline tasks/oss.yml]: Why client?

File: tasks/rdma.yml (hunk 11 above)
Comment (markgoddard) [reply tasks/rdma.yml]: Still not included?

Comment (brtknr) [reply tasks/rdma.yml]: I know, I am not sure if we need it for some other configurations but this is not required for OpenHPC, hence I am reluctant to remove it.

File: tasks/oss.yml (hunk 10 above)
Comment (brtknr) [reply tasks/oss.yml]: Its been reintroduced.

File: defaults/main.yml [hunk 14]
```diff
@@ -20,21 +18,24 @@ beegfs_host_mgmt: ""
...
 
 # File system type
 beegfs_fstype: "xfs"
 
 # Network interfaces in order of preference
-# (leaving empty prefers infiniband and RDMA devices first)
+# (leaving empty means InfiniBand & RDMA enabled devices are preferred)
 beegfs_interfaces: []
+
+beegfs_state: present
```
[the remark this answers is not in the document]
Comment (brtknr) [reply defaults/main.yml]: Done.
